core/frontend/src/extension/Extension.ts

```diff
@@ -34,23 +34,54 @@ export interface ExtensionManifest {
    * e.g "./lib/main.js"
    */
   readonly main: string;
-  /** List of activation events this Extension supports. */
-  readonly activationEvents: ActivationEvent[];
-}
-
-/** @alpha */
-export interface BuildExtensionManifest extends ExtensionManifest {
   /** Only valid when the Extension is loaded at build-time.
    *
    * Defines the main ES module that will be imported
    */
-  readonly module: string;
+  readonly module?: string;
+  /** List of activation events this Extension supports. */
+  readonly activationEvents: ActivationEvent[] | string[];
+}
+
+/**
+ * Methods to provide an extension's content (local function vs remote function)
+ * @alpha
+ */
+interface ExtensionContentProvider {
+  /**
+   * A local function that is the starting execution point for the Extension.
+   * Typically used with locally installed Extensions
+   */
+  readonly main?: ResolveFunc;
+  /**
+   * The url for an endpoint that responds with a Javascript file that contains a default function that is the starting execution point for the Extension.
+   * Typically used with remote Extensions
+   */
+  readonly jsUrl?: string;
 }
 
-/** Describes an Extension that has already been downloaded and has a location files can be easily executed.
+/**
+ * A "ready to use" Extension (contains a manifest object).
+ * Will be used as the type for in-memory extensions in the ExtensionAdmin
  * @alpha
-*/
-export interface LocalExtensionProps {
+ */
+export interface Extension extends ExtensionContentProvider {
   readonly manifest: ExtensionManifest;
-  readonly mainFunc?: ResolveFunc;
+}
+
+/**
+ * Properties that are required to construct an Extension.
+ * Is the parameter type for the "ExtensionAdmin.addExtension" method, which is the preferred method for consumers to provide/register an Extension
+ */
+export interface ExtensionProvider extends ExtensionContentProvider {
```

Comment:
Would the use of @aruniverse's XOR type be useful here instead of all these optional parameters?
Something like `type ExtensionProvider = XOR<LocalExtensionProvider, RemoteExtensionProvider>` since it does not make sense to have both properties

Reply:
I was thinking of maybe moving the xor type into core\bentley\src\UtilityTypes.ts

Reply:
where is it now?

Reply:
Sitting in a branch for a [pr](https://github.com/iTwin/viewer/pull/149) at the moment, https://github.com/iTwin/viewer/blob/types-next/packages/modules/viewer-react/src/types.ts#L30-L33

Reply:
One of the goals of this PR is to blur the lines between local and remote extensions. For example, an extension should be able to provide its javascript at build time but still provide its manifest remotely if it so chooses. I don't believe that we should care to force a method to provide the bundle based on how a user provides the manifest and vice versa.

That said, it does make sense to make the 2 methods to provide the js and the 2 methods to provide the manifest OR. I can update those and swap for the new type if/when added

Reply:
changed to a type with xor properties